File: backend/src/models/conversation_key_backup.py
"""ConversationKeyBackup model — password-recoverable per-conversation message
key backup (mirrors FR-054's identity-key recovery, extended to the 1:1
conversation symmetric key).

The per-conversation AES-256-GCM message key (`conversationCrypto.ts`) is
derived once, client-side, from an ML-KEM-768 shared secret. The *recipient*
side can always re-derive it later by decapsulating the KEM ciphertext carried
in the conversation's first message. The *initiator* side cannot: their copy of
the derived key exists only in `localStorage`, generated once from a random
KEM encapsulation with no ciphertext to redo decapsulation against. Clearing
browser storage therefore permanently loses that user's access to the
conversation's history unless the key itself is backed up somewhere.

This table stores exactly that backup: the message key wrapped (encrypted)
under the same password-derived wrapping key as the user's identity keypair
(FR-054), AAD-bound to the conversation id. Opaque to the server — it stores
and relays ciphertext but never decrypts (the wrapping key is derived
client-side from the password and never sent to the backend).

One row per (conversation, user): each participant backs up their own view of
the key independently.
"""

import uuid
from datetime import UTC, datetime

from sqlalchemy import DateTime, ForeignKey, LargeBinary, String, UniqueConstraint
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column

from src.core.database import Base


class ConversationKeyBackup(Base):
    __tablename__ = "conversation_key_backups"
    __table_args__ = (
        UniqueConstraint("conversation_id", "user_id", name="uq_conversation_key_backup"),
    )

    id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4)
    conversation_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("conversations.id", ondelete="CASCADE"),
        nullable=False,
        index=True,
    )
    user_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True
    )
    # Opaque to the server: AES-256-GCM(wrapKey, messageKey, AAD=conversation_id).
    wrapped_key: Mapped[bytes] = mapped_column(LargeBinary, nullable=False)
    wrap_nonce: Mapped[bytes] = mapped_column(LargeBinary, nullable=False)
    wrap_kdf_salt: Mapped[bytes] = mapped_column(LargeBinary, nullable=False)
    wrap_kdf_params: Mapped[str] = mapped_column(String(64), nullable=False)
    wrap_alg: Mapped[str] = mapped_column(String(32), nullable=False)
    created_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), default=lambda: datetime.now(UTC)
    )
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True),
        default=lambda: datetime.now(UTC),
        onupdate=lambda: datetime.now(UTC),
    )

File: frontend/src/crypto/conversationKeyBackup.ts
/**
 * Conversation-key backup — extends FR-054's password-recoverable identity to
 * the 1:1 conversation symmetric key (see `backend/src/models/conversation_key_backup.py`
 * for the server-side rationale).
 *
 * The recipient side of a 1:1 conversation can always re-derive its message key
 * later by decapsulating the KEM ciphertext embedded in the conversation's
 * first message — their KEM private key is itself recoverable via the
 * password-wrapped identity. The INITIATOR side cannot: their copy of the
 * derived key is a one-time local secret with no ciphertext to redo
 * decapsulation against. Wrapping that key under the same password-derived
 * `wrapKey` as the identity (see `identityWrap.ts`) and pushing the opaque
 * blob to the server closes that gap — any browser that knows the password
 * can recover it, exactly like the identity itself.
 */

import { aes256GcmMessageCipher } from "@/crypto/providers/ciphers";
import { base64ToBytes, bytesToBase64 } from "@/crypto/bytes";
import type { ConversationKeyBackupResponse, PutConversationKeyBackupRequest } from "@/types/messaging";

/** Wrap a conversation's message key under `wrapKey`, AAD-bound to the
 * conversation id (substitution resistance — a backup blob from one
 * conversation can't be replayed as another's). */
export async function wrapConversationKey(
  wrapKey: Uint8Array,
  wrapKdfSalt: Uint8Array,
  wrapKdfParams: string,
  messageKey: Uint8Array,
  conversationId: string,
): Promise<PutConversationKeyBackupRequest> {
  const aad = new TextEncoder().encode(conversationId);
  const { ciphertext, nonce } = await aes256GcmMessageCipher.encrypt(
    wrapKey,
    messageKey,
    aad,
  );
  return {
    wrapped_key: bytesToBase64(ciphertext),
    wrap_nonce: bytesToBase64(nonce),
    wrap_kdf_salt: bytesToBase64(wrapKdfSalt),
    wrap_kdf_params: wrapKdfParams,
    wrap_alg: "aes-256-gcm",
  };
}

/** Unwrap a conversation's message key backup fetched from the server. Throws
 * if `wrapKey` is wrong or the backup was written for a different conversation
 * id (AAD mismatch fails the Poly1305 tag). */
export async function unwrapConversationKey(
  wrapKey: Uint8Array,
  backup: ConversationKeyBackupResponse,
  conversationId: string,
): Promise<Uint8Array> {
  const aad = new TextEncoder().encode(conversationId);
  return aes256GcmMessageCipher.decrypt(
    wrapKey,
    { ciphertext: base64ToBytes(backup.wrapped_key), nonce: base64ToBytes(backup.wrap_nonce) },
    aad,
  );
}
